packages/phone-number-privacy/combiner/src/bls/bls-cryptography-client.ts

```diff
@@ -12,50 +12,95 @@ function flattenSigsArray(sigs: Uint8Array[]) {
+    this.unverifiedSignatures.push(serviceResponse)
   }
 
   /**
    * Returns true if the number of valid signatures is enough to perform a combination
    */
-  public hasSufficientVerifiedSignatures(): boolean {
+  public hasSufficientSignatures(): boolean {
     const threshold = config.thresholdSignature.threshold
-    return this.verifiedSignatures.length >= threshold
+    return this.totalSignatures >= threshold
   }
 
   /*
    * Computes the BLS signature for the blinded phone number.
+   * Throws an exception if not enough valid signatures
+   * and drops the invalid signature for future requests using this instance
    */
-  public async combinePartialBlindedSignatures(): Promise<string> {
+  public async combinePartialBlindedSignatures(
+    blindedMessage: string,
+    logger?: Logger
+  ): Promise<string> {
+    logger = logger ?? rootLogger
     const threshold = config.thresholdSignature.threshold
-    if (!this.hasSufficientVerifiedSignatures()) {
-      const err = new Error(
-        `${ErrorMessage.NOT_ENOUGH_PARTIAL_SIGNATURES} ${this.verifiedSignatures.length}/${threshold}`
+    if (!this.hasSufficientSignatures()) {
+      const error = new Error(
+        `${ErrorMessage.NOT_ENOUGH_PARTIAL_SIGNATURES} ${this.totalSignatures}/${threshold}`
+      )
+      logger.error(error)
+      throw error
+    }
+    // Optimistically attempt to combine unverified signatures
+    // If combination fails, iterate through each signature and remove invalid ones
+    // We do this since signature verification incurs higher latencies
+    try {
+      const result = threshold_bls.combine(threshold, this.allSignatures)
+      return Buffer.from(result).toString('base64')
+    } catch (error) {
+      logger.error(error)
+      // Verify each signature and remove invalid ones
+      // This logging will help us troubleshoot which signers are having issues
+      const verifySigReqs = this.unverifiedSignatures.map((unverifiedSignature) => {
+        return this.verifySignature(blindedMessage, unverifiedSignature, logger!)
+      })
+      await Promise.all(verifySigReqs)
+      this.clearUnverifiedSignatures()
+    }
+    throw new Error(ErrorMessage.NOT_ENOUGH_PARTIAL_SIGNATURES)
+  }
+
+  private async verifySignature(
+    blindedMessage: string,
+    unverifiedSignature: ServicePartialSignature,
+    logger: Logger
+  ) {
+    const sigBuffer = Buffer.from(unverifiedSignature.signature, 'base64')
+    if (await this.isValidSignature(sigBuffer, blindedMessage)) {
+      // We move it to the verified set so that we don't need to re-verify in the future
+      this.verifiedSignatures.push(unverifiedSignature)
+    } else {
+      logger.error({ url: unverifiedSignature.url }, ErrorMessage.VERIFY_PARITAL_SIGNATURE_ERROR)
+    }
+  }
+
+  private clearUnverifiedSignatures() {
+    this.unverifiedSignatures = []
+  }
+
+  private async isValidSignature(signature: Buffer, blindedMessage: string) {
```

Comment:
Is this actually async?

Reply:
Good call. I forgot that we had talked about `threshold_bls.partialVerifyBlindSignature` not actually being async.